lightning/src/ln/channelmanager.rs

```diff
@@ -4102,15 +4106,35 @@ where
 	/// whether or not the payment was successful.
 	///
 	/// [timer tick]: Self::timer_tick_occurred
-	pub fn send_payment_for_bolt12_invoice(&self, invoice: &Bolt12Invoice) -> Result<(), Bolt12PaymentError> {
-		let secp_ctx = &self.secp_ctx;
-		let expanded_key = &self.inbound_payment_key;
-		match invoice.verify(expanded_key, secp_ctx) {
+	pub fn send_payment_for_bolt12_invoice(
+		&self, invoice: &Bolt12Invoice, context: &OffersContext,
+	) -> Result<(), Bolt12PaymentError> {
+		match self.verify_bolt12_invoice(invoice, context) {
 			Ok(payment_id) => self.send_payment_for_verified_bolt12_invoice(invoice, payment_id),
 			Err(()) => Err(Bolt12PaymentError::UnexpectedInvoice),
 		}
 	}
 
+	fn verify_bolt12_invoice(
+		&self, invoice: &Bolt12Invoice, context: &OffersContext,
+	) -> Result<PaymentId, ()> {
+		let secp_ctx = &self.secp_ctx;
+		let expanded_key = &self.inbound_payment_key;
+
+		match context {
+			OffersContext::Unknown {} if invoice.is_for_refund_without_paths() => {
```

Comment:
Are we planning to also stop using `OffersContext::Unknown {}` when creating a reply path for an invoice we're sending to a `Refund`? I thought that change would be in #3192 but don't see it atm. Trying to figure out if this code is going to go away given it only applies to refunds and we seem ok with breaking refunds in this release.

Reply:
> Are we planning to also stop using `OffersContext::Unknown {}` when creating a reply path for an invoice we're sending to a `Refund`? I thought that change would be in #3192 but don't see it atm.

Yeah, looks like I need to update that to use `OffersContext::InboundPayment`. Thanks for catching that!

> Trying to figure out if this code is going to go away given it only applies to refunds and we seem ok with breaking refunds in this release.

This code here will stay the same since the context is from the refund's blinded path -- or `OffersContext::Unknown` if the refund doesn't have any blinded paths. The case that you are thinking of -- the reply path used when creating an invoice _for_ a refund -- would be seen when handling an `InvoiceError`.

Reply:
Ah okay. To me it reads like using an `Option` would capture the state better than having the `Unknown` variant. To me `OffersContext` is context we provide back to ourselves in our blinded paths, and `Unknown` sounds like we got an unknown context in our blinded path, vs no context at all because there was no blinded path.

Reply:
Yeah, I think we wanted to avoid having an `Option` throughout the creation code. But we could remove `OffersContext::Unknown` once we no longer create any paths with it.

Reply:
But if we're avoiding an `Option`, won't `::Unknown` continue to be used when receiving a message to an offer/refund without paths? i.e. here https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/onion_message/messenger.rs#L1516

Reply:
Ah okay, thanks for that clarification.